src/api.ts

```diff
@@ -0,0 +1,559 @@
+> extends BaseApiIntegration {
+  private readonly request: FunctionDecl;
+  private readonly response: FunctionDecl;
+  private readonly errors: { [statusCode: number]: FunctionDecl };
+
+  public constructor(props: Props) {
+    super();
+    this.request = validateFunctionDecl(props.request);
+    this.response = validateFunctionDecl(props.response);
+    this.errors = Object.fromEntries(
+      Object.entries(props.errors).map(([k, v]) => [k, validateFunctionDecl(v)])
+    );
+  }
+
+  public addMethod(
+    httpMethod: HttpMethod,
+    resource: aws_apigateway.Resource
+  ): void {
+    const [requestTemplate, integration] = toVTL(this.request, "request");
+    const [responseTemplate] = toVTL(this.response, "response");
+
+    const errorResponses: aws_apigateway.IntegrationResponse[] = Object.entries(
+      this.errors
+    ).map(([statusCode, fn]) => {
+      const [template] = toVTL(fn, "response");
+      return {
+        statusCode: statusCode,
+        selectionPattern: `^${statusCode}$`,
+        responseTemplates: {
+          "application/json": template,
+        },
+      };
+    });
+
+    const integrationResponses: aws_apigateway.IntegrationResponse[] = [
+      {
+        statusCode: "200",
+        responseTemplates: {
+          "application/json": responseTemplate,
+        },
+      },
+      ...errorResponses,
+    ];
+
+    // TODO: resource is not the right scope, prevents adding 2 methods to the resource
+    // because of the IAM roles created
+    // should `this` be a Method?
+    const apiGwIntegration = integration!.apiGWVtl.makeIntegration(
+      resource,
+      requestTemplate,
+      integrationResponses
+    );
+
+    const methodResponses = [
+      { statusCode: "200" },
+      ...Object.keys(this.errors).map((statusCode) => ({
+        statusCode,
+      })),
+    ];
+
+    resource.addMethod(httpMethod, apiGwIntegration, {
+      methodResponses,
+    });
+  }
+}
+
+/**
+ * Simple VTL interpreter for a FunctionDecl. The interpreter in vtl.ts
+ * is based on the AppSync VTL engine which has much more flexibility than
+ * the API Gateway VTL engine. In particular it has a toJson utility function
+ * which means the VTL can just create an object in memory and then toJson it
+ * the end. Here we need to manually output the JSON which is how VTL is
+ * typically meant to be used.
+ *
+ * For now, only Literals and references to the template input are supported.
+ * It is definitely possible to support more, but we will start with just
+ * small core and support more syntax carefully over time.
+ *
+ * @param node Function to interpret.
+ * @param template Whether we are creating a request or response mapping template.
+ */
+export function toVTL(
+  node: FunctionDecl,
+  location: "request" | "response"
+): [string, IntegrationImpl<any> | undefined] {
+  // TODO: polish these error messages and put them into error-codes.ts
+  if (node.body.statements.length !== 1) {
```

Comment:
we will want to support basic statements before the return stmt e.g. `const x = 1; return fn({x})`. these will need to be interpreted differently than the return statement and will be more like the appsync variant of VTL

waiting until I get to unit tests to do this